### ISSUE TITLE
train_seg out of memory with a batch size of 8

### ISSUE THREAD

**Issue by zhiqi-li** [TencentARC/T2I-Adapter#38]
HI, the paper reports the model is trained with a batch size of 8 on 32G V100, but I got out of memory with the default settings in train_seg.py.  When I set the batch size to 4, the memory is about 27G, so I am slightly confused about this problem because the learning rate maybe needs to adjust when I set a different batch size from yours.

**Comment by zhiqi-li** [comment]
I also want to ask if the training log of train_seg.py can be provided as a reference.

**Comment by MC-E** [comment]
We use distributed training on four V100 GPUs, each with the batchsize being 2.

**Comment by zhiqi-li** [comment]
For the coco stuff dataset, one image corresponds to several captions, what I want to ask is whether the amount of data in one epoch is equal to the number of captions or the number of images. If it is the number of captions, one epoch has more than 600k data, and it seems that it needs much longer time than 2 days for 10 epochs.

**Comment by blackmagicianZ** [comment]
> For the coco stuff dataset, one image corresponds to several captions, what I want to ask is whether the amount of data in one epoch is equal to the number of captions or the number of images. If it is the number of captions, one epoch has more than 600k data, and it seems that it needs much longer time than 2 days for 10 epochs.

The same issue, I trained on 8 V100, with a total batch size of 8x2, The time of one epoch was about 10 hours.

**Comment by MC-E** [comment]
In coco dataset, each image contains 5 captions. The current open source code is that these 5 captions appear in an epoch. The training procedure we use is to randomly select a caption for each image. The adapter converges quickly with training.

**Comment by wanghao14** [comment]
@zhiqi-li @blackmagicianZ Hi, have you successfully trained the model to achieve results close to those in the paper?

**Comment by wanghao14** [comment]
@MC-E Hi, I am training your code on semantic segmentation map. Do you remember how many epochs  it takes to see the convergence of adapter in this condition?

**Comment by MERONAL** [comment]
@MC-E @zhiqi-li Hi, I can't complete training on multiple GPUs, each training only uses one GPU, how do I start multi-GPU training?

**Comment by enkaranfiles** [comment]
> > For the coco stuff dataset, one image corresponds to several captions, what I want to ask is whether the amount of data in one epoch is equal to the number of captions or the number of images. If it is the number of captions, one epoch has more than 600k data, and it seems that it needs much longer time than 2 days for 10 epochs.
> 
> The same issue, I trained on 8 V100, with a total batch size of 8x2, The time of one epoch was about 10 hours.

What was your one GPU size, I am curious? Is there any-way to work on this repo(in SD branch - excluding XL part), 8-16 GB V100 GPU? I am kind of angry to Google, there is no available resource for A100 right now, all my work stuck in the machine!

**Comment by shoutOutYangJie** [comment]
how mush training step can arise Controlable ability？ I found loss is very small at the begin of training.

**Comment by dmmSJTU** [comment]
@wanghao14 Hi, Could you please add your WeChat and ask some questions about training? my email is dmm2020@sjtu.edu.cn

**Comment by dmmSJTU** [comment]
@zhiqi-li Hi, Could you please add your WeChat and ask some questions about training? my email is dmm2020@sjtu.edu.cn.

**Comment by wanghao14** [comment]
> @wanghao14 Hi, Could you please add your WeChat and ask some questions about training? my email is [dmm2020@sjtu.edu.cn](mailto:dmm2020@sjtu.edu.cn)

Post your question here instead of requesting personal contact information.

**Comment by dmmSJTU** [comment]
Hi Wanghao,
https://github.com/TencentARC/T2I-Adapter/blob/16bba674b472121d5a86e3ed6b935f91d516bc74/train_sketch.py#L231 
How do you obtain the mask images of train2017_color? Are you using stuff_train2017_pixelmaps?
Look forward for your reply.

发件人: Wang Hao
发送时间: 2023年12月27日 0:05
收件人: TencentARC/T2I-Adapter
抄送: dmmSJTU; Comment
主题: Re: [TencentARC/T2I-Adapter] train_seg out of memory with a batchsize of 8 (Issue #38)

@wanghao14 Hi, Could you please add your WeChat and ask some questions about training? my email is ***@***.***
Post your question here instead of requesting personal contact information.

**Comment by wanghao14** [comment]
@dmmSJTU Yes, I have used the segmentation map provided by COCO and converted the IDs (pixel class) to RGB values using [this code](https://github.com/PITI-Synthesis/PITI/blob/main/preprocess/preprocess_mask.py).  I have trained an image inpainting model on segmentation condition based on the idea of T2I-Adapter and it works well. 

Hope this could help you. 

You can also refer to:
#22 #25

**Comment by dmmSJTU** [comment]
Thank you. When I used “CUDA_VISIBLE_DEVICES=0,1 python3 -m torch.distributed.launch --nproc_per_node=2  --use-env train_seg.py --ckpt models/sd-v1-4.ckpt --bsize 2”, it produced:

 
Could you help me to solve it?
发件人: Wang Hao
发送时间: 2023年12月27日 15:26
收件人: TencentARC/T2I-Adapter
抄送: dmmSJTU; Mention
主题: Re: [TencentARC/T2I-Adapter] train_seg out of memory with a batchsize of 8 (Issue #38)

@dmmSJTU Yes, I have used the segmentation map provided by COCO and converted the IDs (pixel class) to RGB values using this code. I have trained an image inpainting model on segmentation condition based on the idea of T2I-Adapter and it works well.
Hope this could help you.
You can also refer to:
#14 #22

**Comment by dmmSJTU** [comment]
@wanghao14  @zhiqi-li @MERONAL   Hi, When i train the train_seg.py on single gpu, it produce:
<img width="827" alt="Snipaste_2023-12-28_14-58-55" src="https://github.com/TencentARC/T2I-Adapter/assets/111695222/44812013-0b64-4c36-b2e9-dd2b56e5617f">
Besides, When i use "CUDA_VISIBLE_DEVICES=0,1 python3 -m torch.distributed.launch --nproc_per_node=2  --use-env train_seg.py --ckpt models/sd-v1-4.ckpt --bsize 2" to run on multi-gpu, it produce: 
![123](https://github.com/TencentARC/T2I-Adapter/assets/111695222/78bf17c4-4575-4776-a928-a2c453e0ebb3)

Have you encountered the same problem?
​Look forward to your reply!

**Comment by wanghao14** [comment]
@dmmSJTU 
There appears to be an issue with the initialization of distributed training. Please verify the number of GPUs available in your environment and check whether a command has been included in the code to specify the utilization of a particular graphics card, for example, using 'os.environ['CUDA_VISIBLE_DEVICES'] = '0'."

This issue might not be related to this code.

**Comment by dmmSJTU** [comment]
wanghao 同学你好，可以加你好友问几个问题吗？在这里问比较麻烦，时效性也不太好。    

-----原始邮件-----

发件人: Wang ***@***.***>
目标语言: TencentARC ***@***.***>
抄送: dmmSJTU ***@***.***>; Mention ***@***.***>
日期: 2023年12月28日星期四 15:41 CST
主题: Re: [TencentARC/T2I-Adapter] train_seg out of memory with a batch size of 8 (Issue #38)

 

  @dmmSJTU
There appears to be an issue with the initialization of distributed training. Please verify the number of GPUs available in your environment and check whether a command has been included in the code to specify the utilization of a particular graphics card, for example, using 'os.environ['CUDA_VISIBLE_DEVICES'] = '0'." This issue might not be related to this code. —

**Comment by wanghao14** [comment]
I am sorry, but the question you've asked isn't pertinent to this project. It seems to be related to your personal environment configuration, and I'm not interested in it.
